recipes/api/tests/test_recipe_api.py

```diff
@@ -0,0 +1,214 @@
+from django.test import TestCase
+from django.urls import reverse
+from rest_framework.test import APIClient
+from rest_framework import status
+
+from api.serializers import RecipeSerializer
+from core.models import Recipe, Ingredient
+
+RECIPES_URL = reverse("api:recipes-list")
+
+
+def detail_url(recipe_id):
+    """Return recipe detail URL
+
+    :param recipe_id:
+    :return:
+    """
+    return reverse("api:recipe-detail", args=[recipe_id])
+
+
+def sample_recipe(name, **params):
+    """Helper function to create recipes
+
+    :param name:
+    :return:
+    """
+    ingredients = params.pop("ingredients", None)
+    recipe = Recipe.objects.create(name=name, **params)
+    if ingredients:
+        for name in ingredients:
+            recipe.ingredients.add(Ingredient.objects.create(name=name))
+
+    return recipe
+
+
+class RecipeAPITests(TestCase):
+    """None
+
+    """
+
+    def setUp(self) -> None:
+        self.client = APIClient()
+
+    def test_retrieve_basic_recipe_list(self):
+        """Test retrieving recipes
+
+        :return:
+        """
+        sample_recipe(name="Bolognaise")
+
+        res = self.client.get(RECIPES_URL)
+
+        recipes = Recipe.objects.all().order_by("-name")
+        serializer = RecipeSerializer(recipes, many=True)
+
+        self.assertEqual(status.HTTP_200_OK, res.status_code)
+        self.assertEqual(serializer.data, res.data)
+
+    def test_create_basic_recipe(self):
+        """Test creating recipes
+
+        :return:
+        """
+        payload = {
+            "name": "Bolognaise",
+            "description": "Put it in the oven",
+            "ingredients": [
+                {"name": "dough"}, {"name": "cheese"}, {"name": "tomato"}
+            ]
+        }
+
+        res = self.client.post(RECIPES_URL, payload, format="json")
+
+        self.assertEquals(status.HTTP_201_CREATED, res.status_code)
+        recipe = Recipe.objects.get(name=payload["name"])
+        serializer = RecipeSerializer(recipe, many=False)
+
+        self.assertEqual(serializer.data, res.data)
+
+    def test_create_basic_recipe_no_ingredients(self):
```

Comment:
Often helpful to include expected outcome in test name to be clear, e.g. `test_create_basic_recipe_with_no_ingredients_fails`

Reply:
OK! I guess I just always think that the outcome might change over time and then you might have test names that don't correspond with the current outcome. I.e. additional maintenance!